tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,767 @@
+	bytesWritten += MintMaxtSize
+
+	cdm.writeBuf[bytesWritten] = byte(chk.Encoding())
+	bytesWritten += ChunkEncodingSize
+
+	n := binary.PutUvarint(cdm.writeBuf[bytesWritten:], uint64(len(chk.Bytes())))
+	bytesWritten += n
+
+	if err := cdm.writeAndAppendToCRC32(cdm.writeBuf[:bytesWritten]); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.writeAndAppendToCRC32(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	if err := cdm.writeCRC32(); err != nil {
+		return 0, err
+	}
+
+	if maxt > cdm.curFileMaxt {
+		cdm.curFileMaxt = maxt
+	}
+
+	if mint < cdm.curFileMint {
+		cdm.curFileMint = mint
+	}
+
+	cdm.chunkBuffer.put(chkRef, chk)
+
+	if len(chk.Bytes())+MaxHeadChunkMetaSize >= writeBufferSize {
+		// The chunk was bigger than the buffer itself.
+		// Flushing to not keep partial chunks in buffer.
+		if err := cdm.flushBuffer(); err != nil {
+			return 0, err
+		}
+	}
+
+	return chkRef, nil
+}
+
+func chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+// shouldCutNewFile decides the cutting of a new file based on time and size retention.
+// Size retention: because depending on the system architecture, there is a limit on how big of a file we can m-map.
+// Time retention: so that we can delete old chunks with some time guarantee in low load environments.
+func (cdm *ChunkDiskMapper) shouldCutNewFile(chunkSize int, maxt int64) bool {
+	return cdm.curFileNumBytes == 0 || // First head chunk file.
+		(maxt-cdm.curFileMint > cdm.maxFileMs && cdm.curFileNumBytes > HeadChunkFileHeaderSize) || // Time duration reached for the existing file.
+		cdm.curFileNumBytes+int64(chunkSize+MaxHeadChunkMetaSize) > MaxHeadChunkFileSize // Exceeds the max head chunk file size.
+}
+
+func (cdm *ChunkDiskMapper) cut(mint int64) (returnErr error) {
+	// Sync current tail to disk and close.
+	if err := cdm.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(cdm.dir, MagicHeadChunks, headChunksFormatV1, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// The file should not be closed if there is no error,
+		// its kept open in the ChunkDiskMapper.
+		if returnErr != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(returnErr)
+			merr.Add(f.Close())
+			returnErr = merr.Err()
+		}
+	}()
+
+	cdm.size += cdm.curFileNumBytes
+	atomic.StoreInt64(&cdm.curFileNumBytes, int64(n))
+
+	if cdm.curFile != nil {
+		// Open it again with the new size.
+		newTailFile, err := fileutil.OpenMmapFile(cdm.curFile.Name())
+		if err != nil {
+			return err
+		}
+		cdm.readPathMtx.Lock()
+		oldFileCloser := cdm.closers[cdm.curFileSequence]
+		cdm.closers[cdm.curFileSequence] = newTailFile
+		cdm.mmappedChunkFiles[cdm.curFileSequence] = &mmappedChunkFile{
+			byteSlice: realByteSlice(newTailFile.Bytes()),
+			maxt:      cdm.curFileMaxt,
+		}
+		cdm.readPathMtx.Unlock()
+		// Closing the last mmapped file.
+		if err := oldFileCloser.Close(); err != nil {
```

Comment:
I was guessing this would be confusing :)

So, `oldFileCloser` and `cdm.curFile` are 2 different things.
1. The closers have closers to the m-mapped files which have a read-only file open within them. Hence `oldFileCloser` is for the m-mapped file.
2. `cdm.curFile` is the only write-only file open, and, it is closed in `finalizeCurFile`.

This means the last file is open in 2 places - m-mapping and writing. So I guess no changes here and the comments should be there I guess to explicitly say its for the m-mapped file.

Reply:
but in this case it is the same file no?

Reply:
Same file, but opened in 2 places, so we should close both the object. The files objects are not shared as one of them is read-only.

Reply:
To other reviewers: After some offline discussion, we came to conclusion that we won't require reloading the mmapped file again and we can reuse the already mmapped file with a larger initial size.